ctk/common/src/main/kotlin/org/codice/compliance/verification/core/SubjectComparisonVerifier.kt

```diff
@@ -0,0 +1,267 @@
+ * the License, or any later version.
+ *
+ * <p>This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+ * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details. A copy of the GNU Lesser General Public
+ * License is distributed along with this program and can be found at
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ */
+package org.codice.compliance.verification.core
+
+import org.codice.compliance.SAMLComplianceException
+import org.codice.compliance.SAMLCore_3_3_4_b1
+import org.codice.compliance.SAMLCore_3_3_4_b2
+import org.codice.compliance.SAMLCore_3_3_4_c1
+import org.codice.compliance.SAMLCore_3_3_4_c2
+import org.codice.compliance.SAMLCore_3_4_1_4_a
+import org.codice.compliance.SAMLCore_3_4_1_4_a1
+import org.codice.compliance.SAMLProfiles_4_1_4_2_e
+import org.codice.compliance.attributeList
+import org.codice.compliance.attributeText
+import org.codice.compliance.children
+import org.codice.compliance.recursiveChildren
+import org.codice.compliance.utils.TestCommon.Companion.FORMAT
+import org.codice.compliance.utils.TestCommon.Companion.SUBJECT
+import org.w3c.dom.Node
+
+/**
+ * Verifies the request subjects **strongly match** the response subject according to the Core
+ * document sections 3.3.4 & 3.4.1.4. (verifySubjectsMatchSSO)
+ *
+ * Verifies the response subjects refer to the same principal according to the Profiles document
+ * section 4.1.4.2. (verifySubjectsMatchAuthnRequest)
+ *
+ * @param request Optional AuthnRequest {@code Node}. Should be passed when calling the
+ * verifySubjectsMatchAuthnRequest method.
+ * @param response Response {@code Node}.
+ */
+class SubjectComparisonVerifier(private val request: Node? = null, private val response: Node) {
+
+    companion object {
+        private const val UNSPECIFIED_URI =
+                "urn:oasis:names:tc:SAML:2.0:nameid-format:unspecified"
+        private const val EMPTY_STRING = ""
+    }
+
+    /**
+     * This function verifies that all of the Subjects in the Response refer to the same principal.
+     *
+     * NOTE: This method does not actually resolve who the principal of the subject is. It only
+     * tests that if two Subject identifiers have the same format, their content is the same. In
+     * order to fully test this function we need to resolve the Subjects to a principal.
+     *
+     * NOTE: This method does not actually resolve who the principal of the subject is. It only
+     * compares the given Subject values to each other.
+     */
+    fun verifySubjectsMatchSSO() {
+        val subjectList = response.children("Assertion")
+                .flatMap { it.children(SUBJECT) }
+                .toList()
+
+        subjectList.forEachIndexed { index, outerSubject ->
+            subjectList.subList(index + 1, subjectList.size).forEach { innerSubject ->
```

Comment:
❗️ I thought we decided you didn't need two nested loops here and that you could just do one pass.

Reply:
✏️ While a little less efficient, for the size of our sets this is a lot simpler and gets us out of the business of managing combinatorial math.
```kotlin
        val subjectSet = response.children("Assertion")
                .flatMap { it.children(SUBJECT) }
                .toSet()
        Sets.combinations(subjectSet, 2).forEach {
            verifyIdContentsMatchSSO(it.first(), it.last())
        }
```
Can also inline the `subjectSet` to make it even more concise.

Reply:
✏️ The verifyIdContentsMatchSSO check appears to be transitive, so I believe the following should accomplish the same thing and avoid checking all combinations. 
```
subjectList.forEachIndexed { i, outerSubject ->
            val innerSubject = subjectList.get(i + 1)
            verifyIdContentsMatchSSO(outerSubject, innerSubject)
}
```